File: sorting/quicksort.py

def partition(arr,low,high):
    i=low+1
    j=high
    pivot=arr[low]
    while(i<=j):
        #find element greater than pivot
        while(arr[i]<=pivot):
            i+=1
        
        #find element lesser than piviot
        while(arr[j]>pivot):
            j-=1

       #swap that greater than and lesser than pivot element
        if(i<j):
            arr[i],arr[j]=arr[j],arr[i]

    arr[low],arr[j]=arr[j],arr[low] #placing piviot element in its position of sorted array
    return j   # j is the index of piviot element in sorted array


def quicksort(arr,low,high):
    n=len(arr)
    
    if(low<high):
        parititionIndex=partition(arr,low,high)  #gives index of pivot element in sorted array

        quicksort(arr,low,parititionIndex-1)   # sorting the sub-array of elements less than the pivot(left sub array)
        quicksort(arr,parititionIndex+1,high)  # sorting the sub-array of elements greater than the pivot(right sub array)
    return arr

# arr=[5,81,55,74,0,4,60,74,5,2,9]
arr=[1,2,3,4,5,6,7,8]
sorted_arr=quicksort(arr,low=0,high=len(arr)-1)
print(sorted_arr)

File: sorting/quick_sort.c
#include <stdio.h>

// Function to partition the array and return the pivot index
int partition(int arr[], int low, int high) {
    int pivot = arr[low];
    int i = low + 1;
    int j = high;

    while (i <= j) {
        // Find element greater than pivot
        while (i <= high && arr[i] <= pivot) {
            i++;
        }

        // Find element lesser than pivot
        while (arr[j] > pivot) {
            j--;
        }

        // Swap elements greater than and lesser than pivot
        if (i < j) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
        else {
            break;
        }
    }

    // Swap pivot with element at index j
    int temp = arr[low];
    arr[low] = arr[j];
    arr[j] = temp;

    return j; // Return the pivot index
}

// Function to perform quicksort
void quicksort(int arr[], int low, int high) {
    if (low < high) {
        int partitionIndex = partition(arr, low, high);

        // Recursively sort elements before and after partition
        quicksort(arr, low, partitionIndex - 1);
        quicksort(arr, partitionIndex + 1, high);
    }
}

int main() {
    int arr[] = {15, 20, 55, 8, 15};
    int n = sizeof(arr) / sizeof(arr[0]);

    quicksort(arr, 0, n - 1);

    printf("Sorted array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}
